### PR TITLE
various 0.7 fixes

### PR DESCRIPTION
**Pull request by ExpandingMan**
I've cleaned up a lot of this simple stuff, but this package is in need of serious attention from someone intimately familiar with it for 0.7.  Many tests are passing without warning or error, but in particular when it gets to `missing`s things go awry, and it looks like some of the problems may involve design decisions.

### CODE REVIEW

File: src/CategoricalArrays.jl [hunk 1]
```diff
@@ -13,11 +13,14 @@ module CategoricalArrays
 
     using Compat
     using Reexport
+
+    # TODO: cannot @reexport in conditional, the below should be removed when 0.6 is deprecated
     @reexport using Missings
 
-    if VERSION >= v"0.7.0-DEV.3052"
+    if VERSION >= v"0.7.0-"
```
Comment (nalimilan) [inline src/CategoricalArrays.jl]: Why this change? If the version isn't correct, it should be fixed rather than removed.

File: src/CategoricalArrays.jl [hunk 2]
```diff
@@ -13,11 +13,14 @@ module CategoricalArrays
 
     using Compat
     using Reexport
+
+    # TODO: cannot @reexport in conditional, the below should be removed when 0.6 is deprecated
```
Comment (nalimilan) [inline src/CategoricalArrays.jl]: Even inside `@static if VERSION > ...`?

File: src/subarray.jl [hunk 3]
```diff
@@ -6,12 +6,18 @@ isordered(sa::SubArray{T,N,P}) where {T,N,P<:CategoricalArray} = isordered(paren
 levels!(sa::SubArray{T,N,P}, newlevels::Vector) where {T,N,P<:CategoricalArray} =
     levels!(parent(sa), levels)
 
+if VERSION ≥ v"0.7.0-"
```
Comment (nalimilan) [inline src/subarray.jl]: That's `0.7.0-DEV.3020`. Also, given that Femtocleaner will remove 0.6-only blocks for us in a few weeks, I'd rather duplicate the full `unique` body instead of defining `_subarray_indices`, which Femtocleaner won't remove automatically for us.

File: test/08_string.jl [hunk 4]
```diff
@@ -4,6 +4,13 @@ using Compat.Test
 using Compat.Unicode
 using CategoricalArrays
 
+if VERSION ≥ v"0.7.0-"
```
Comment (nalimilan) [inline test/08_string.jl]: That's `0.7.0-DEV.2915`. Below, I guess you don't need to `import` the function at all, we can just define our own function since `LinAlg.normalize` isn't used here. I wonder whether it would also work in DataFrames.

File: test/08_string.jl [hunk 5]
```diff
@@ -138,10 +145,10 @@ using CategoricalArrays
     @test match(r"af", v2, 2).offset === 2
     @test match(r"af", v2, 2, UInt32(0)).offset === 2
 
-    @test matchall(r"af", v1) == []
-    @test matchall(r"af", v2) == ["af"]
-    if VERSION > v"0.7.0-DEV.3526"
-        @test matchall(r"af", v2, overlap=true) == ["af"]
+    @test collect((m.match for m ∈ eachmatch(r"af", v1))) == []
+    @test collect((m.match for m ∈ eachmatch(r"af", v2))) == ["af"]
+    if VERSION > v"0.7.0-"
```
Comment (nalimilan) [inline test/08_string.jl]: Why remove the exact version?

File: test/10_isless.jl [hunk 6]
```diff
@@ -253,61 +253,61 @@ end
...
-    @test isless(v1, v1) === false
-    @test isless(v1, v2) === true
-    @test isless(v1, v3) === true
-    @test isless(v2, v1) === false
-    @test isless(v2, v2) === false
-    @test isless(v2, v3) === true
-    @test isless(v3, v1) === false
-    @test isless(v3, v2) === false
-    @test isless(v3, v3) === false
+    pool1 = CategoricalPool(["a", "b", "c"])
```
Comment (nalimilan) [inline test/10_isless.jl]: Please avoid renaming variables just to silence warnings, as requested in DataFrames.

File: src/CategoricalArrays.jl (hunk 1 above)
Comment (ExpandingMan) [reply src/CategoricalArrays.jl]: ... I'm pretty sure there was a reason but honestly I forget.  Will revert.

File: src/CategoricalArrays.jl (hunk 2 above)
Comment (ExpandingMan) [reply src/CategoricalArrays.jl]: hm, don't know, will experiment

File: src/subarray.jl (hunk 3 above)
Comment (ExpandingMan) [reply src/subarray.jl]: ok, will change.

File: test/08_string.jl (hunk 4 above)
Comment (ExpandingMan) [reply test/08_string.jl]: Sure, I can do that if you want, but I didn't want to risk it in case someone had a commit that used `LinearAlgebra.normalize` (unlikely, I know, but I figured why not account for the possibility)

File: test/08_string.jl (hunk 5 above)
Comment (ExpandingMan) [reply test/08_string.jl]: Again, now I forget what happened.  I'm pretty sure one of these caused a problem and then I got overzealous and removed all of them.

File: test/10_isless.jl (hunk 6 above)
Comment (ExpandingMan) [reply test/10_isless.jl]: will revert

File: test/08_string.jl (hunk 4 above)
Comment (nalimilan) [reply test/08_string.jl]: Hopefully if such a thing happens, it will be after we drop 0.6 support. :-p

File: test/10_isless.jl (hunk 6 above)
Comment (nalimilan) [reply test/10_isless.jl]: I guess you can call them `s1` or `cs1`, etc., if you really want to kill those warnings.

Comment (ExpandingMan) [reply test/10_isless.jl]: Sorry, to seem like I am being such a pain just for the sake of these warnings.  I can appreciate that they seem like a triviality, and for any aspect for the package other than updating it that is true.  However, having just been through the process of updating to 0.7, please believe me when I say that they caused me significant confusion.  It can be very difficult to distinguish between intended behavior and bugs unless the tests are clean.  This is coming from someone who's somewhat familiar with DataFrames, someone coming in cold might find it impossible to do such an update, even though the vast majority of the changes that ultimately wound up happening here should only require a working knowledge of core Julia.  It's easy to walk away from tests at the end of the day and forget about the warnings, but every time a PR happens it's a potential source of confusion.

File: src/CategoricalArrays.jl (hunk 2 above)
Comment (ExpandingMan) [reply src/CategoricalArrays.jl]: apparently not.  I think what's happening is that when `@reexport` macro gets called at compile time, it checks for `Missings` and it does this regardless of whether it's in a conditional or not.  let me know if you think of an alternative...

Comment (nalimilan) [reply src/CategoricalArrays.jl]: Concretely, it is a problem to reexport Missings.jl on 0.7? It should work fine there, and we can keep it that way until Femtocleaner removes the block.

File: test/10_isless.jl (hunk 6 above)
Comment (nalimilan) [reply test/10_isless.jl]: Yes, these are rough times. Hopefully this annoying situation won't happen again after 1.0 is released.

File: src/CategoricalArrays.jl (hunk 2 above)
Comment (ExpandingMan) [reply src/CategoricalArrays.jl]: Well, I'm stumped.  It doesn't seem to work in the conditional statement no matter what.  Doesn't matter if `Missing`s is installed, doesn't even matter if you already have it imported into the context where `CategoricalArrays` is imported.  As far as I can tell right now there is aboslutely no way to get this inside a conditional statement.

Comment (nalimilan) [reply src/CategoricalArrays.jl]: Let's make it unconditional for now, that's not a big deal since AFAIK all exported objects are also in Base now.

**Comment by nalimilan** [comment]
See https://github.com/JuliaData/CategoricalArrays.jl/pull/132 for one failure which triggered more errors down the line.

**Comment by ExpandingMan** [comment]
Any more requested changes?  Like I said there are still failing tests on 0.7, but I think this is all I'll do for this PR.  It's a start...

**Comment by nalimilan** [comment]
Thanks!
